plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveNodePartitioningProvider.java

```diff
@@ -101,7 +108,22 @@ public ToIntFunction<ConnectorSplit> getSplitBucketFunction(
             ConnectorSession session,
             ConnectorPartitioningHandle partitioningHandle)
     {
-        return value -> ((HiveSplit) value).getReadBucketNumber()
-                .orElseThrow(() -> new IllegalArgumentException("Bucket number not set in split"));
+        final HivePartitioningHandle handle = (HivePartitioningHandle) partitioningHandle;
+        if (handle.getPartitions().isEmpty()) {
+            return value -> ((HiveSplit) value).getReadBucketNumber()
+                    .orElseThrow(() -> new IllegalArgumentException("Bucket number not set in split"));
+        }
+        List<String> partitions = handle.getPartitions();
+        ImmutableMap.Builder<String, Integer> partitionToIndexBuilder = ImmutableMap.builder();
+        for (int i = 0; i < partitions.size(); i++) {
+            partitionToIndexBuilder.put(partitions.get(i), i);
+        }
+        final ImmutableMap<String, Integer> partitionToIndex = partitionToIndexBuilder.buildOrThrow();
```

Comment:
Don't use `final` on local variables. Declare as the interface type `Map` rather than the implementation `ImmutableMap`.

Reply:
Done.